tortoise/state.go

```diff
@@ -5,78 +5,242 @@ import (
+	layer.blocks = append(layer.blocks, block)
+	s.blockRefs[block.id] = block
+	s.updateRefHeight(layer, block)
+}
+
+func (s *state) updateRefHeight(layer *layerInfo, block *blockInfo) {
+	if layer.verifying.referenceHeight == 0 && layer.lid.After(s.evicted) {
+		layer.verifying.referenceHeight = s.layer(layer.lid.Sub(1)).verifying.referenceHeight
+	}
+	if block.height <= s.referenceHeight[block.layer.GetEpoch()] &&
+		block.height > layer.verifying.referenceHeight {
+		layer.verifying.referenceHeight = block.height
+	}
+}
+
+type (
+	blockVote struct {
+		*blockInfo
+		vote sign
+	}
+
+	baseInfo struct {
+		id    types.BallotID
+		layer types.LayerID
+	}
+
+	conditions struct {
+		baseGood bool
+		// is ballot consistent with local opinion within [base.layer, layer)
+		consistent bool
+		// set after comparing with local beacon
+		badBeacon bool
+		// any exceptions before base.layer
+		votesBeforeBase bool
+	}
+
+	ballotInfo struct {
+		id         types.BallotID
+		layer      types.LayerID
+		base       baseInfo
+		height     uint64
+		weight     weight
+		beacon     types.Beacon
+		votes      votes
+		conditions conditions
+	}
+)
+
+func (b *ballotInfo) good() bool {
+	return b.canBeGood() && b.conditions.baseGood
+}
+
+func (b *ballotInfo) canBeGood() bool {
+	return !b.conditions.badBeacon && !b.conditions.votesBeforeBase && b.conditions.consistent
+}
+
+type votes struct {
+	tail *layerVote
+}
+
+func (v *votes) append(lv *layerVote) {
+	if v.tail == nil {
+		v.tail = lv
+	} else {
+		v.tail = v.tail.append(lv)
+	}
+}
+
+func (v *votes) update(from types.LayerID, diff map[types.LayerID]map[types.BlockID]sign) votes {
+	if v.tail == nil {
+		return votes{}
+	}
+	return votes{tail: v.tail.update(from, diff)}
+}
+
+// cutBefore cuts all pointers to votes before the specified layer.
+func (v *votes) cutBefore(lid types.LayerID) {
+	for current := v.tail; current != nil; current = current.prev {
+		prev := current.prev
+		if prev != nil && prev.lid.Before(lid) {
+			current.prev = nil
+			return
+		}
+		if current.lid.Before(lid) {
+			v.tail = nil
+			return
+		}
+	}
+}
+
+func (v *votes) find(lid types.LayerID, bid types.BlockID) sign {
+	for current := v.tail; current != nil; current = current.prev {
+		if current.lid == lid {
+			for _, block := range current.blocks {
+				if block.id == bid {
+					return block.vote
+				}
+			}
+		}
+	}
+	return abstain
+}
+
+type layerVote struct {
+	*layerInfo
+	vote   sign
+	blocks []blockVote
+
+	prev *layerVote
+}
+
+func (l *layerVote) copy() *layerVote {
+	return &layerVote{
+		layerInfo: l.layerInfo,
+		vote:      l.vote,
+		blocks:    l.blocks,
+		prev:      l.prev,
+	}
+}
+
+func (l *layerVote) append(lv *layerVote) *layerVote {
+	lv.prev = l
+	return lv
+}
+
+func (l *layerVote) update(from types.LayerID, diff map[types.LayerID]map[types.BlockID]sign) *layerVote {
+	if l.lid.Before(from) {
+		return l
+	}
+	l = l.copy()
+	if l.prev != nil {
+		l.prev = l.prev.update(from, diff)
```

Comment:
not a fan of recursion. i know it's more elegant but wouldn't this be a source of memory overhead compared to an iterative approach.

as a reader, it's more intuitive to iterate the diff from `from` to `ballot.layer.Sub(1)`, and have `next` instead `prev`
the code as such, the main part of `layerVote.update()` is very similar to the last part of `decodeException` (// add new opinions after the base layer)

Reply:
i realize that there will be additional overhead, but it is not about elegance. list supports iteration only in one direction, for example:
1 <- 2 <- 3 <- 4
will allow to iterate from 4 to 1. what i need to do is to find the first layer where opinion changed and cope list from that point

for example child ballot changes opinion on layer 2, so the list should become

1 <- 2' <- 3' <- 4'

Reply:
`next` doesn't work for several reasons, but the most significant is:
 
given the same list 1 -> 2 -> 3 -> 4
and for example i want to append new opinion -> 5

i can't simply 1 -> 2 -> 3 -> 4 -> 5 because it modifies 4, but if i copy 4 i also need to copy 3 and so on

Reply:
yeah... thinking more about it, `next` doesn't work indeed.